cardano-testnet/src/Testnet/Ping.hs

```diff
@@ -0,0 +1,145 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# OPTIONS_GHC -Wno-partial-fields #-}
+
+module Testnet.Ping
+  ( pingClient
+  , TestnetMagic
+  ) where
+
+import           Control.Exception
+import           Control.Monad (when)
+import           Control.Monad.Class.MonadTime.SI (Time)
+import           Control.Tracer (nullTracer)
+import           Data.Word (Word32)
+import           Network.Mux.Bearer (MakeBearer (..), makeSocketBearer)
+import           Network.Mux.Timeout (TimeoutFn, withTimeoutSerial)
+import           Network.Mux.Types (MiniProtocolDir (InitiatorDir), MiniProtocolNum (..),
+                   MuxBearer (read, write), MuxSDU (..),
+                   MuxSDUHeader (MuxSDUHeader, mhDir, mhLength, mhNum, mhTimestamp),
+                   RemoteClockModel (RemoteClockModel))
+import           Network.Socket (AddrInfo, StructLinger (..))
+
+import           Cardano.Api (Error (..))
+import           Cardano.Api.Pretty
+import           Cardano.Network.Ping (HandshakeFailure, NodeVersion (..), handshakeDec,
+                   handshakeReq, isSameVersionAndMagic, supportedNodeToClientVersions)
+import qualified Codec.CBOR.Read as CBOR
+import qualified Control.Monad.Class.MonadTimer.SI as MT
+import qualified Data.ByteString.Lazy as LBS
+import qualified Data.List as L
+import qualified Hedgehog.Extras.Stock.IO.Network.Sprocket as IO
+import qualified Network.Socket as Socket
+import           Prettyprinter
+
+type TestnetMagic = Word32
+
+handshakeNum ::  MiniProtocolNum
+handshakeNum = MiniProtocolNum 0
+
+sduTimeout :: MT.DiffTime
+sduTimeout = 30
+
+doHandshakeQuery :: Bool
+doHandshakeQuery = True
+
+pingClient :: TestnetMagic -> IO.Sprocket -> IO (Either PingClientError ())
+pingClient networkMagic sprocket = bracket
```

Comment:
This is a gutted-out `pingClient` from cardano-ping. One alternative to this would be to refactor `cardano-ping` to not print directly to stdout and use it here.